From: Vlad Yasevich <vyasevic@redhat.com>
Date: Thu, 27 Mar 2014 22:14:46 -0400
Subject: [PATCH 24/32] qlge: Do not propaged vlan tag offloads to vlans
Patch-mainline: v3.14
Git-commit: f6d1ac4b5f15f57929fe0fa283b3a45dfec717a0
References: bsc#909361 FATE#317547

qlge driver turns off NETIF_F_HW_CTAG_FILTER, but forgets to
turn off HW_CTAG_TX and HW_CTAG_RX on vlan devices.  With the
current settings, q-in-q will only generate a single vlan header.
Remember to mask off CTAG_TX and CTAG_RX features in vlan_features.

CC: Shahed Shaikh <shahed.shaikh@qlogic.com>
CC: Jitendra Kalsaria <jitendra.kalsaria@qlogic.com>
CC: Ron Mercer <ron.mercer@qlogic.com>
Signed-off-by: Vlad Yasevich <vyasevic@redhat.com>
Acked-by: Jitendra Kalsaria <jitendra.kalsaria@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Gary Ching-Pang Lin <glin@suse.com>
---
 drivers/net/ethernet/qlogic/qlge/qlge_main.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/drivers/net/ethernet/qlogic/qlge/qlge_main.c
+++ b/drivers/net/ethernet/qlogic/qlge/qlge_main.c
@@ -4744,7 +4744,9 @@ static int qlge_probe(struct pci_dev *pd
 	ndev->features = ndev->hw_features;
 	ndev->vlan_features = ndev->hw_features;
 	/* vlan gets same features (except vlan filter) */
-	ndev->vlan_features &= ~NETIF_F_HW_VLAN_FILTER;
+	ndev->vlan_features &= ~(NETIF_F_HW_VLAN_FILTER |
+				 NETIF_F_HW_VLAN_TX |
+				 NETIF_F_HW_VLAN_RX);
 
 	if (test_bit(QL_DMA64, &qdev->flags))
 		ndev->features |= NETIF_F_HIGHDMA;
